src/jit/simdcodegenxarch.cpp

```diff
@@ -730,9 +730,10 @@ void CodeGen::genSIMDScalarMove(
 
 void CodeGen::genSIMDZero(var_types targetType, var_types baseType, regNumber targetReg)
 {
-    // pxor reg, reg
-    instruction ins = getOpForSIMDIntrinsic(SIMDIntrinsicBitwiseXor, baseType);
-    inst_RV_RV(ins, targetReg, targetReg, targetType, emitActualTypeSize(targetType));
+    // We just use `INS_xorps` instead of `getOpForSIMDIntrinsic(SIMDIntrinsicBitwiseXor, baseType)`
+    // since `genSIMDZero` is also used elsewhere and we no longer have a gurantee that AVX and AVX2
+    // will both be enabled
+    inst_RV_RV(INS_xorps, targetReg, targetReg, targetType, emitActualTypeSize(targetType));
```

Comment:
The comment should be clarified. I would replace "is also use elsewhere" with "is used for both the System.Numerics.Vector types as well as the HardwareIntrinsic types". and add that vpxor is not available for 256-bit operands on AVX.
Also, would it be preferable to us this only in the case where `targetType` is `TYP_SIMD32`?

Reply:
> Also, would it be preferable to us this only in the case where targetType is TYP_SIMD32?

@fiigii might be able to comment (or know who can) on whether there is any partial stalls. IIRC, some instructions have a stall when you switch between floating-point and integer operations on the same register; however, given that these are dependency-chain breaking instructions and have no execution latency, I'm not sure if that's the case here. If there is no partial stall, I don't think it is necessarily worthwhile to have the difference.

> The comment should be clarified

I'll update

Reply:
The type is kind of irrelevant here, even if you use TYP_SIMD16 you'll still going to zero out the entire ymm register.

Reply:
I think always using `INS_xorps` is okay, `xorps` can be handled by the register renaming and not reached CPU backend. Additionally, with SSE encoding, `xorps` is one byte shorter than other instructions.

Reply:
> some instructions have a stall when you switch between floating-point and integer operations on the same register

`xorps reg, reg` is handled by register renaming, before reaching the INT/FP execution unit, on any reasonable CPU. And if you're dealing with an unreasonable CPU, well, you may as well discover that `xorps` is anyway executed on the INT unit so it's not like it really matters.

Reply:
@mikedn, are you suggesting that we just emit the appropriate instruction (`INS_xorps`, `INS_xorpd`, or `INS_pxor`), as we do today, but always do it for `TYP_SIMD16`? 

If I understood correctly, this should work since, under the VEX-encoding it will also zero the upper bits; and under non-VEX it won't matter since TYP_SIMD32 is invalid...

Reply:
> are you suggesting that we just emit the appropriate instruction (INS_xorps, INS_xorpd, or INS_pxor), as we do today, but always do it for TYP_SIMD16?

If I understand correctly the problem this PR is trying to fix, always using `TYP_SIMD16` would be an option if you want to keep emitting `pxor`.

But then I don't see any good reason to emit `pxor`. `xorps` was more commonly used before VEX because it had shorter encoding. With VEX the encoding size is the same but then there's still little reason to use `pxor` for integer vectors and/or complicate things even more by doing it only for `TYP_SIMD32`. 

So it's always `xorps`, with the proper type (just because I don't see any reason not to use the proper type with `xorps`). And it's not a small, targeted fix, it's just the normal implementation.

Reply:
> So it's always xorps, with the proper type (just because I don't see any reason not to use the proper type with xorps).

That sounds reasonable to me. I can update the comment appropriately.

> And it's not a small, targeted fix, it's just the normal implementation

It's still a somewhat targeted fix. After this change we still have an issue where we don't have anything that will catch/prevent this type of mistake.

That is, the emitter currently just decides to use the VEX-encoding if VEX is supported at all, which leads to problems if you are on some older hardware (like we have in our pool of Mac machines) that only supports AVX. Ideally we would have something (possibly just an assert) that will catch when we try to emitting an AVX2 instruction on hardware that only has AVX support (i.e. we try to emit an SSE2 integer instruction with TYP_SIMD32).

Reply:
>  Ideally we would have something (possibly just an assert) that will catch when we try to emitting an AVX2 instruction on hardware that only has AVX support (i.e. we try to emit an SSE2 integer instruction with TYP_SIMD32).

Yeah, an assert would be good if it's not too much trouble. This kind of bug sucks because it's difficult to test, you can turn off JIT's support for AVX2 but it's more difficult to turn off hardware support for AVX2.

Reply:
Fixed the comment. Building everything locally on my i5-3317 (which is a CPU that has AVX, but no AVX2 support) now so I can validate the change.